Upgrade project to support EPiServer 12

Almaroo opened the pull request:
**General**
- [ ] support .NET 6
- [ ] refactor code to meet .NET 6 guidelines

**Implementation**
- [ ] move logic from Helpers to TagBuilders
- [ ] enable creating support for different provider (by default ImageSharp)

**Tests**
- [ ] add tests for `CssHelpers`
- [ ] add tests for `PictureTagBuilder`
- [ ] add tests for `SourceTagBuilder`
- [ ] add tests for `ImageSharpResizedUrlGenerator`

Discuss whether we should change `IImage` and `IResponsiveImage` hierarchy
Discuss which enchantments should be introduced

Almaroo commented on EpiResponsivePicture/GeneratorProfiles/ResizedUrlGeneratorBase.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+        protected UrlBuilder Builder;
+
+        private readonly Queue<Func<int, PictureSource, FocalPoint, (string Key, string Value)>> CustomQueryRegistrations = new();
```
Replace `Func<int, PictureSource, FocalPoint, (string Key, string Value)>` with `using RegisterQueryFunc = Func<int, PictureSource, FocalPoint, (string Key, string Value)>`, `Queue<RegisterQueryFunc>`

Almaroo commented on EpiResponsivePicture/GeneratorProfiles/ResizedUrlGeneratorBase.cs:
```diff
@@ -0,0 +1,51 @@
+        protected abstract (string Key, string Value) QualityQuery(PictureSource source);
+        protected abstract (string Key, string Value) FormatQuery(ResizedImageFormat format);
+
+        protected void RegisterCustomQuery(Func<int, PictureSource, FocalPoint, (string Key, string Value)> registration) =>
```
I guess whole `PictureProfile` should be passed in here, to allow access to `MaxImageDimension`, `Format` in provider specific generator

Almaroo commented on EpiResponsivePicture/ResizedImage/UrlBuilderExtensions.cs:
```diff
@@ -19,6 +19,15 @@ public static UrlBuilder Add(this UrlBuilder target, string key, string value)
+        public static UrlBuilder Add(this UrlBuilder target, (string Key, string Value) queryTuple)
+        {
+            if (target == null)
+                throw new ArgumentNullException(nameof(target));
```
Replace with `Guard`

tomaszmadeyski replied:
or use new syntax `ArgumentNullException.ThrowIfNull(target)`

Almaroo commented on EpiResponsivePicture/TagBuilders/PictureTagBuilder.cs:
```diff
@@ -0,0 +1,155 @@
+
+        private void GetImgAltText(IImage image)
+        {
+            altImgText = image.Description ?? string.Empty;
```
Change variable name to `altImageText` for consistency

Shall alt fallback to some meaningful value like 'image'? On the one hand it is recommended to have alt tag on the other hardcoded value will not be locale specific. Maybe mark `Description` property in `IImage` as `[Required]`?

Almaroo commented on EpiResponsivePicture/ResizedImage/CssHelpers.cs:
```diff
@@ -0,0 +1,62 @@
+
+        private static string BuildSize(ValueTuple<double, Unit> width) => $"{width.Item1.ToString(CultureInfo.InvariantCulture)}{width.Item2.AsString()}";
+        
+        internal static string AsString(this Unit unit) => unit switch
```
Not a big fan of this, however enum's `ToString` cannot be overriden and custom converter seemed like an overkill

Almaroo commented on EpiResponsivePicture/ResizedImage/Property/FocalPoint.cs:
```diff
@@ -14,6 +14,8 @@ public FocalPoint(double x, double y)
+
         public override string ToString()
         {
             return $"{X.ToString(CultureInfo.InvariantCulture)}|{Y.ToString(CultureInfo.InvariantCulture)}";
```
Add rounding to 3rd decimal place

Almaroo commented on EpiResponsivePicture/TagBuilders/PictureTagBuilder.cs:
```diff
@@ -0,0 +1,155 @@
+        private void GetPictureData()
+        {
+            var imageFound = ServiceLocator.Current.GetInstance<IContentLoader>().TryGet<IContentData>(pictureContentReference,
+                new LoaderOptions { LanguageLoaderOption.FallbackWithMaster() }, out var content) && content is IImage;
```
Could there be a real world scenario where someone decides to go with interfaces instead of a base class and makes use only of `IImage` without implementing `IResponsiveImage`? 

According to readme:

> If you don't like the inheritance, alternatively you can implement interfaces: `IImage` (for resizing & proper alt loading), and `IResponsiveImage` (for focal point and proper cropping).

I am unsure how resizing works in such a scenario? Via `AspectRatio`? What is the expected behaviour when content is of type `IImage` and `AspectRatio` is not set?

Almaroo replied:
Currently solved with common `IImageWithWidthAndSize` interface

tomaszmadeyski commented on EpiResponsivePicture.Tests/EpiResponsivePicture.Tests.csproj:
```diff
@@ -1,25 +1,25 @@
-    <PackageReference Include="Moq" Version="4.5.28" />
-    <PackageReference Include="NUnit" Version="3.13.1" />
+    <PackageReference Include="Castle.Core" Version="5.0.0" />
+    <PackageReference Include="EPiServer.CMS.Core" Version="12.6.0" />
```
It seems that the first version that has support for .net 6 is `12.4.0` so I think we should specify this as `Version="[12.4.0, 13.0)"`. By this we're saying "we don't care what version you have unless it is at least 12.4 and not 13"

tomaszmadeyski replied:
(same goes for episerver.framework)

tomaszmadeyski commented on EpiResponsivePicture.Tests/EpiResponsivePicture.Tests.csproj:
```diff
@@ -1,25 +1,25 @@
 
   <ItemGroup>
-    <Folder Include="obj\Release\netcoreapp3.1" />
+    <Folder Include="obj\Release\net5.0" />
```
I think we don't need this entry?

tomaszmadeyski commented on EpiResponsivePicture/Configuration/Constants.cs:
```diff
@@ -0,0 +1,7 @@
+    {
+        internal const string ModuleName = "Forte.EpiResponsivePicture";
+    }
+}
```
I see that new files are missing new line at the end of the file. You can update Rider's settings so it automatically adds it for you

tomaszmadeyski commented on EpiResponsivePicture/Configuration/Constants.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Forte.EpiResponsivePicture.Configuration
+{
```
a tip: in .net 6 you don't need to use braces for namespace

tomaszmadeyski commented on EpiResponsivePicture/Configuration/ResponsivePictureModule.cs:
```diff
@@ -3,8 +3,9 @@
+using Forte.EpiResponsivePicture.ResizedImage;
 
-namespace Forte.EpiResponsivePicture.ResizedImage
+namespace Forte.EpiResponsivePicture.Configuration
```
I think we can clear this a bit and move logic from this class to `AddForteEpiResponsivePicture` method. You can check an example [here](https://github.com/fortedigital/EpiEasyEvents/blob/master/src/EpiEasyEvents/ServiceCollectionExtensions.cs) how to use EventsRegistry

(it's probably a matter of opinion but I feel that using this `[InitializableModule]` does not fit to .net core and `.AddMyCustomExtension()` kind of doing it

tomaszmadeyski commented on EpiResponsivePicture/EpiResponsivePicture.csproj:
```diff
@@ -19,67 +21,44 @@
-    <PackageReference Include="EPiServer.CMS.UI.Core" Version="11.11.0" />
-    <PackageReference Include="Microsoft.AspNet.Mvc" Version="5.2.3" />
+    <PackageReference Include="Baaijte.Optimizely.ImageSharp.Web" Version="2.0.0" />
+    <PackageReference Include="EPiServer.CMS.AspNetCore" Version="12.6.0" />
```
same thing with version specification as in Tests case (range instead of concrete version number)

tomaszmadeyski commented on EpiResponsivePicture/EpiResponsivePicture.nuspec:
```diff
@@ -0,0 +1,31 @@
+    </metadata>
+    <files>
+        <file src="bin/Release/EpiResponsivePicture.dll" target="lib/net6.0" />
+        <file src="bin/Release/Forte.EpiResponsivePicture.zip" target="contentFiles/any/any/modules/_protected/Forte.EpiResponsivePicture" />
```
hmm, what's that `any/any` in the path? 🤔

tomaszmadeyski commented on EpiResponsivePicture/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+        private static void ConfigureModule(IServiceCollection services)
+        {
+            services.AddSingleton<ImagePublishingEventHandler>();
+            services.AddTransient<IResizedUrlGenerator, ImageSharpResizedUrlGenerator>();
```
since we're assuming that someone will want to replace implementation then maybe it's better to use `services.TryAddTransient` syntax?

tomaszmadeyski commented on EpiResponsivePicture/ResizedImage/ResizedImageFormat.cs:
```diff
@@ -6,8 +8,11 @@ public enum ResizedImageFormat
+        [Obsolete($"Use {nameof(Jpeg)} instead")]
+        Jpg = Jpeg,
+        Jpeg = 1,
         Png,
```
I think it's worth to set int values for all enum values

tomaszmadeyski commented on EpiResponsivePicture/TagBuilders/PictureTagBuilder.cs:
```diff
@@ -0,0 +1,154 @@
+            return element;
+        }
+        
+        private void GetPictureData()
```
IMHO all these Get* methods look odd when they return nothing (void). These seem to be rather Set than Get, don't you agree?

tomaszmadeyski commented on EpiResponsivePicture/TagBuilders/ISourceTagBuilder.cs:
```diff
@@ -0,0 +1,16 @@
+        ISourceTagBuilder WithFocalPoint(FocalPoint point);
+        ISourceTagBuilder WithResizedImageFormat(ResizedImageFormat format);
+        TagBuilder Build();
+        ISourceTagBuilder Clear();
```
you added Clear to `ISourceTagBuilder`. Shouldn't we add, for the sake of consistency, add Clear method to `IPictureTagBuilder` as well?

tomaszmadeyski commented on EpiResponsivePicture/TagBuilders/SourceTagBuilder.cs:
```diff
@@ -0,0 +1,88 @@
+
+        public ISourceTagBuilder Clear()
+        {
+            element = new TagBuilder("source");
```
hmm, I think this method might be a bit misleading - it says "Clear" and I expected that it would clear whole builder with all attributes, meaning:

```
var builder = SourceTagBuilder.Create()
.WithImageUrl("someUrl"); 

builder.Clear();
```

after that I would expect that ImageUrl is empty and this is not the case here

tomaszmadeyski commented on EpiResponsivePicture/TagBuilders/SourceTagBuilder.cs:
```diff
@@ -0,0 +1,88 @@
+        
+        private IEnumerable<string> GetSourceSets() => pictureSource.AllowedWidths.Select(BuildSize);  
+
+        private string BuildSize(int width)
```
BuildWidth?

Almaroo commented on EpiResponsivePicture/TagBuilders/PictureTagBuilder.cs:
```diff
@@ -0,0 +1,154 @@
+            {
+                if (responsiveImage is null || pictureProfileSource.TargetAspectRatio is null)
+                {
+                    pictureProfileSource.TargetAspectRatio = AspectRatio.Default; // 4:3 
```
Note: hidden bug in here. PictureProfile is being modified while rendering a picture.

tomaszmadeyski commented on EpiResponsivePicture/ResizedImage/PictureProfile.cs:
```diff
@@ -1,10 +1,11 @@
-}
\ No newline at end of file
+    public int DefaultWidth { get; init; }
+    public int MaxImageDimension { get; init; } = 3200; // Default max served by ImageResizer TODO check how it is in ImageSharp
```
did you check this value (3200) for ImageSharp?

Almaroo replied:
I was not able to find any constraints on dimensions

tomaszmadeyski replied:
so let's remove this TODO comment then

kpkozak commented:
Will it be .Net 6 only? It makes impossible to update for any other users that are still stuck on Epi v11 and .net framework. If so, maybe it is worth considering just renaming the package (and also follow Epi->Optimizely transition). Especially that it kinda changes the idea (from overgrown expansion over Valdis Iljuconoks' package to something independent from image processing tool + default ImageSharp support)

Also, provide some documentation on how to use this software with ImageSharp (as it does not seem to have dependency over it). Remove [EPiServerBlobReader](https://github.com/valdisiljuconoks/ImageResizer.Plugins.EPiServerBlobReader) dependency. 
It may then be worth to have base/core package + providers package(s) that adds needed dependencies on ImageSharp.

So we have: 
- `EpiResponsivePicture` (or `EpiResponsivePicture.Core`) for core functionality
- `EpiResponsivePicture.ImageSharp` for support of ImageSharp: dependency on ImageSharp, registering handlers + Url provider
- Alternatively - other `EpiResponsivePicture.XXXX` - if we want to still have support for another resizing tool.

Does it make sense @tomaszmadeyski @Almaroo ?

tomaszmadeyski commented:
@kpkozak 

> Will it be .Net 6 only? It makes impossible to update for any other users that are still stuck on Epi v11 and .net framework. If so, maybe it is worth considering just renaming the package (and also follow Epi->Optimizely transition). Especially that it kinda changes the idea (from overgrown expansion over Valdis Iljuconoks' package to something independent from image processing tool + default ImageSharp support)

It is planned only for .net5/.net6 but since .net5 has reached end of life we kept .net6 only. The approach here is similar to other EpiServer nugets we maintain in Forte ([AzureSearch ](https://github.com/fortedigital/EpiServer.AzureSearch) of [EpiEasyEvents](https://github.com/fortedigital/EpiEasyEvents)) - meaning master branch is targetting new framework and new major version of nuget is released. We keep up-to-date master as a `v1` branch and every change commited to this branch will release new version of nuget but with previous major.

On example of AzureSearch:
master -> nuget v3 
v2 -> nuget v2

It is similar to ie EpiServer.Find - v13 is targetting EpiServer v11 (hence .net framework) and v14 is targetting EpiServer v12 (hence .net5/6)

> Also, provide some documentation on how to use this software

+1 from me

> It may then be worth to have base/core package + providers package(s) that adds needed dependencies on ImageSharp.

We've been considering splitting core & imageSharp into different nugets but for the sake of simplicity of usage. I assume 90% will just use ImageSharp (which BTW is a dependency in this project) - for them it's only installing single nuget and single `.Add*()` and `.Use*()` clause in Startup class. For the rest of the user's we're leaving an opportunity to provide own implementation of imageProcessor (and possibly create own nugets as an extensions)

kpkozak commented:
I didn't have time to fully look through and analyze all the changes, so sorry for stupid questions and/or not precise comments.

Here's my feedback after just a brief look:
- How the caching works now? Do we need to configure ImageSharp caching in any way? It is not mentioned in docs at all
- There is no solution in readme on how to use library **without** a need to derive from base type. This is the case when we're trying to use library in existing project, with image content types defined, migrating from other etc or when simply developers have aversion to deriving wit no real reason. Should be covered in docs and library too.
- ImagePointEditor is directly thrown and commited into the project - while it should be the EpiResponsivePicture dependency, installed together with the package.
- No migration guide between v1 & v2. Please add some explanation article, link it.
- Migrating.md covers EpiCommonUtils -> EpiResponsivePicture v1. Let's state this in the article, don't mislead it can be about v1 -> v2.
- What's the point of `IImageWithWidthAndHeight`? Name of interface sounds weird, does not suggest usage

Other small comments:
- CSS helpers look to be outside the scope of the library. They do not perform any real function. There is no point in having them at all, and they cover only simplest, single-condition queries which make them not very useful. Writing more complex media queries, etc. makes necessary to write such things by yourself anyway. In addition, it covers things that the picture specification does not support (for example, %, which is not supported in srcset and will not work). This makes all misleading to potential users
- ImageCropper.cs - there is a comment that always there must be crop due to issue in ImageResizer. Dropping ImageResizer makes that obsolete, we no longer need to have crop parameter every time, so we need FocalPoint.Center even less.
- BMP format has no use in web, no need to add that. However, if you use it in your project, let's leave it
- ImmutableArray vs Array  in picture profiles. This is not major issue, just a bit of simplicity (using just an array used to be a bit more cochise). But why specifically ImmutableArray? Actually, everything what we need is just iterating over, so we could live with some interface like IReadOnlyCollection, or even IEnumerable<T>.
- Why AspectRatio.Default is 4:3? There is no point in having/suggesting any default as every use case is different. Also, there is no point in chosing this particular value (why not 16:9, or 1:1 then etc.)
- Default max served by ImageResizer is a default max for Picture profile. What is max for ImageSharp? Do we want to leave 3200 as max here? If so, remove comment that says about ImageResizer as we're no longer using it.

Almaroo commented:
@kpkozak @tomaszmadeyski 
I have updated the readme, yet still, the project is lacking a decent migration guide

> * What's the point of IImageWithWidthAndHeight? Name of interface sounds weird, does not suggest usage

It tries to somehow fill the gap between IImage and IResponsiveImage. We let the users to inherit from IImage only, while the whole functionality of the library was based on IResponsiveImage (which led to unexpected behaviours, as only IResponsiveImage had its dimensions calculated on upload)

> * ImagePointEditor is directly thrown and commited into the project - while it should be the EpiResponsivePicture dependency, installed together with the package.

I moved it in here due to a bug that forced me to modify library's code. For a reason that's unknown to me _FocalPoint_ value was no longer passed to dojo as a string but rather as an object. That resulted in having lots of console errors and focal point indicator not working correctly.

> * ImmutableArray vs Array in picture profiles. This is not major issue, just a bit of simplicity (using just an array used to be a bit more cochise). But why specifically ImmutableArray? Actually, everything what we need is just iterating over, so we could live with some interface like IReadOnlyCollection, or even IEnumerable.

I think the main reason why I went with ImmutableArray was the fact that I have accidentally introduced a bug that could potentially lead to mutating existing PictureProfiles. And ImmutableArray somehow felt most natural to me. If you think it is an overkill I could totally go with IReadOnlyCollection instead  

> * Why AspectRatio.Default is 4:3? There is no point in having/suggesting any default as every use case is different. Also, there is no point in chosing this particular value (why not 16:9, or 1:1 then etc.)

No particular reason, thought it could be useful. Can be removed.
